src/seqn/trans.lisp

```diff
@@ -0,0 +1,399 @@
+                                                                       cadrobj))
+                                      ((= carobjnum 1) (infix-creation :+
+                                                                       ifzero
+                                                                       ifone))
+                                      (t output))))
+                             ((typep cadrobj 'vamp:constant)
+                              (let ((cadrobjnum (vamp:const cadrobj)))
+                                (cond ((= cadrobjnum 0) (infix-creation :*
+                                                                        ifzero
+                                                                        carobj))
+                                      ((= cadrobjnum 1) (infix-creation :+
+                                                                        (infix-creation :*
+                                                                                        ifzero
+                                                                                        carobj)
+                                                                        car))
+                                      (t  output))))
+                             (t output))))
+         (prod-list left
+                    right)))))
+
+(defmethod to-vampir ((obj shift-front) inputs constraints)
+  "Takes the MCADR entry and moves it upward leaving everything
+else fixed. Note that we have to be careful as inputs will have 0es
+removed already and hence we cannot count as usual"
+  (declare (ignore constraints))
+  (append (cons (nth (1- (mcadr obj))
+                     inputs)
+                (subseq inputs 0 (1- (mcadr obj))))
+          (subseq inputs (mcadr obj))))
+
+(defmethod to-vampir ((obj zero-bit) inputs constraints)
+  (declare (ignore inputs constraints))
+  (list (vamp:make-constant :const 0)))
+
+(defmethod to-vampir ((obj one-bit) inputs constraints)
+  (declare (ignore inputs  constraints))
+  (list (vamp:make-constant :const 1)))
+
+(defmethod to-vampir ((obj seqn-add) inputs constraints)
+  (declare (ignore constraints))
+  (let ((car (car inputs))
+        (cadr (cadr inputs))
+        (mcar (mcar obj)))
+    (if (and (typep car 'vamp:constant)
+             (typep cadr 'vamp:constant))
+        (let ((plus (+ (vamp:const car)
+                       (vamp:const cadr))))
+          (if (> (expt 2 mcar) plus)
+              (vamp:make-constant :const plus)
+              (error "Range Exceeded")))
+        (list (geb.vampir:plus-range (vamp:make-constant :const mcar)
+                                     car
+                                     cadr)))))
+
+(defmethod to-vampir ((obj seqn-subtract) inputs constraints)
+  (declare (ignore constraints))
+  (let ((car (car inputs))
+        (cadr (cadr inputs)))
+    (if (and (typep car 'vamp:constant)
+             (typep cadr 'vamp:constant))
```

Comment:
This pattern is very common, should we make a function that checks if the first two are constants?

Reply:
Good idea!